src/common/multi_file_reader.cpp

```diff
@@ -587,33 +743,123 @@ void MultiFileReader::CreateMapping(const string &file_name,
+	//! If any of these fail, the file can be skipped entirely
+	auto evaluation_result = EvaluateConstantFilters(filters, file_name, global_columns, global_column_ids,
+	                                                 virtual_columns, reader_data, global_to_local);
+	if (evaluation_result.can_skip_file) {
+		return false;
 	}
 
-	reader_data.filter_map.resize(filter_map_size);
+	reader_data.filters = CreateFilters(evaluation_result.remaining_filters, global_to_local);
+	return true;
+}
+
+static unique_ptr<TableFilter> ConvertFilterFromGlobalToLocal(const TableFilter &global_filter,
+                                                              MultiFileIndexMapping &mapping) {
+	auto type = global_filter.filter_type;
+
+	unique_ptr<TableFilter> result;
+	switch (type) {
+	case TableFilterType::CONSTANT_COMPARISON:
+	case TableFilterType::IS_NULL:
+	case TableFilterType::IS_NOT_NULL:
+	case TableFilterType::IN_FILTER:
+		return global_filter.Copy();
+	case TableFilterType::CONJUNCTION_OR: {
+		auto &or_filter = global_filter.Cast<ConjunctionOrFilter>();
+		auto res = make_uniq<ConjunctionOrFilter>();
+		for (auto &it : or_filter.child_filters) {
+			auto child_filter = ConvertFilterFromGlobalToLocal(*it, mapping);
+			if (child_filter) {
+				res->child_filters.push_back(std::move(child_filter));
+			}
+		}
+		return std::move(res);
+	}
+	case TableFilterType::CONJUNCTION_AND: {
+		auto &and_filter = global_filter.Cast<ConjunctionAndFilter>();
+		auto res = make_uniq<ConjunctionAndFilter>();
+		for (auto &it : and_filter.child_filters) {
+			auto child_filter = ConvertFilterFromGlobalToLocal(*it, mapping);
+			if (child_filter) {
+				res->child_filters.push_back(std::move(child_filter));
+			}
+		}
+		return std::move(res);
+	}
+	case TableFilterType::STRUCT_EXTRACT: {
+		auto &struct_filter = global_filter.Cast<StructFilter>();
+		auto &child_filter = struct_filter.child_filter;
+
+		//! FXIME: The previous step should ensure that filters that target fields that are not present in the file are
+		//! evaluated earlier
+		//! For now we will assume the mapping is 1-to-1
+		MultiFileIndexMapping mapping(struct_filter.child_idx);
+		auto new_child_filter = ConvertFilterFromGlobalToLocal(*child_filter, mapping);
+		if (!new_child_filter) {
+			return nullptr;
+		}
+		//! TODO: renaming fields should probably be respected here?
+		auto child_name = struct_filter.child_name;
+		return make_uniq<StructFilter>(mapping.index, child_name, std::move(new_child_filter));
+	}
+	case TableFilterType::OPTIONAL_FILTER: {
+		//! For now we'll skip all the optional filters, as they are not necessary for correctness
+		return nullptr;
+	}
+	case TableFilterType::DYNAMIC_FILTER: {
+		//! FIXME: since this works like BoundParameterExpression,
+		// making a copy of this DynamicFilterData will not function correctly.
+		throw NotImplementedException("DynamicFilter can't be *properly* copied yet");
```

Comment:
Since a dynamic filter can only contain constant filters we can just copy them over as-is